File: sol/127.cpp
#include <queue>
#include <unordered_set>

struct Solution {
    int ladderLength(string &beginWord, string &endWord, vector<string>& wordList) {
        unordered_set<string> wl(wordList.begin(), wordList.end());
        if (!wl.contains(endWord)){
            return 0;
        }
        int dep = 2;
        queue<string> q;
        q.push(beginWord);
        while (!q.empty()){
            int s = q.size();
            while (s-- > 0){
                string w = q.front();
                q.pop();
                for (int i = 0; i < w.size(); ++i){
                    char c = w[i];
                    for (int j = 0; j < 26; ++j){
                        w[i] = 'a' + j;
                        if (w == endWord){
                            return dep;
                        }
                        if (wl.contains(w)){
                            wl.erase(w);
                            q.push(w);
                        }
                    }
                    w[i] = c;
                }
            }
            ++dep;
        }
        return 0;
    }
};

File: sol/127/127.py
from collections import deque

class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        wordList = set(wordList)
        if endWord not in wordList:
            return 0
        dep = 2
        q = deque((beginWord,))
        while (s := len(q)) != 0:
            for _ in range(s):
                w = list(q.popleft())
                for i in range(len(w)):
                    c = w[i]
                    for j in range(26):
                        w[i] = chr(ord('a') + j)
                        ww = ''.join(w)
                        if ww == endWord:
                            return dep
                        if ww in wordList:
                            wordList.remove(ww)
                            q.append(ww)
                    w[i] = c
            dep += 1
        return 0
